### DISCUSSION TITLE
Why RON?

### DISCUSSION THREAD

**Discussion by cjs-axsh** [emilk/egui#3523] [Q&A]
The persistence feature for eframe serializes to RON before saving to file. Why? As far as I can tell, the only benefit of RON is that it makes writing enums by hand easier.  But are you expecting anyone to be editing the persistence files by hand?

Since serde is being used for the serialization, I would prefer to use a format that supports all of serde's features, at least for the full round trips. JSON is probably the most obvious contender in that regard. If it were CBOR or something, I might be able to see a how the trade-offs worked in favor of non-JSON, but with RON I'm not seeing the benefit. 

Why this matters to me specifically is that I have an sdk with a lot of newtypes and enums that RON serializes into a format that it can't deserialize back from, so I can't use the persistence feature with these types as is. My current least-effort workaround is to serialize to JSON and put that into a String that eframe then puts into the RON file. And then on loading, I have to get the string and deserialize manually to instantiate the app. This seems unnecessarily wasteful and inconvenient.

Anyway, if there's some reason that I'm unaware of for using RON, I'd love to learn what it is. Otherwise, is there any chance of changing to a more reliable method of saving app state (e.g. serde + json)?

**Comment by kvark** [comment]
@cjs-axsh have you filed this as an issue to RON repo? This sounds like a bug.

**Comment by cjs-axsh** [comment]
@kvark no, it is well documented in their repo already, including in the [limitations](https://github.com/ron-rs/ron#limitations) section of their readme.

**Reply by kvark** [reply]
My understanding (as well as the original design) was to guarantee a roundtrip as long as you are using serde-Serialize/Deserialize annotated types, i.e. described types. I'd expect the limitations section to only apply to self-describing aspect of RON, i.e. "ron::Value", and that is supported by this last paragraph:
> While RON guarantees roundtrips like Rust -> RON -> Rust for Rust types using non-deserialize_any-based implementations, RON does not yet make any guarantees about roundtrips through ron::Value

I might be missing the critical detail here, and I'd like to understand more. Would you happen to have the example that fails to roundtrip, and which doesn't rely on self-described deserialization?

**Reply by cjs-axsh** [reply]
The critical detail you seem to be missing is in your quoted text:
> using non-deserialize_any-based implementations

This is what I meant when I said that I'd prefer a solution that supports all serde features if serde is instrumental to the implementation. For examples, you can just look at the issues linked in their readme.

edit: I don't mean to brush you off with the examples -- I just haven't spent the effort to figure out exactly what is failing in RON for my case. Since they document several cases that don't work, it doesn't seem worth the time to figure out vs just using something that works. This is why the question I asked here is "Why RON?" rather than "RON's great, let's all try to fix it." as I don't see the benefit of trying to expand RON for eframe usage instead of just using JSON. If this were a feature request issue, I would probably ask for the option to change formats and keep RON as the default for now as it seems like the least breaking way to do it, but for now I was just trying to understand why RON was chosen in the first place.

**Reply by cjs-axsh** [reply]
Also, how does eframe do the roundtrips? Just re-reading your quote, 
> RON does not yet make any guarantees about roundtrips through ron::Value

would mean that there are no guarantees if eframe uses `ron::Value` in the implementation, right? I mean I could look this up if I really wanted to, but maybe since you quoted it you know some of the details about eframe's implementation that you are assuming I also read into? Just to be clear, I haven't looked at all at the implementation and wasn't trying to imply anything about using or not using `ron::Value`s.

**Reply by kvark** [reply]
It doesn't look like eframe is using any of the deserialize_any stuff, based on https://github.com/emilk/egui/blob/f218825d94f8bbaffd6e9bc1119b1073d013f629/crates/eframe/src/epi/mod.rs#L1182-L1206
Looks pretty straightforward, and with this kind of usage RON guarantees correct roundtrip. So I'm still missing something here - how did this fail for you?
And yes, you are correct, you aren't asking "how to fix RON", you are asking a different question. I'm not able to help you with the original question, but I think the failure is worth looking into as well.

**Reply by cjs-axsh** [reply]
> and with this kind of usage RON guarantees correct roundtrip. So I'm still missing something here 

Well, it's good to know that eframe is not explicitly breaking the method it chose, but that does not mean that all data structures passed to eframe will avoid using anything related to deserialize_any.

> how did this fail for you?

I am still not certain and still not ready to spend time figuring it out, but based on the discussion thus far, presumably something in my sdk is using deserialize_any, right? Either that or a bug in RON -- neither of which interest me.

**Reply by cjs-axsh** [reply]
Maybe I should restate the original problem as I see it with more clarity about what is not being supported currently. I think that if eframe is using serde to serialize to a file for persistence, then it should support all serde features. So from the RON limitations section linked above,
- #[serde(tag = "tag")], i.e. internally tagged enums [1](https://github.com/ron-rs/ron#user-content-fn-serde-enum-hack-59fb02d3c6aa3d4ae2c84d8a096b4d39)
- #[serde(tag = "tag", content = "content")], i.e. adjacently tagged enums [1](https://github.com/ron-rs/ron#user-content-fn-serde-enum-hack-59fb02d3c6aa3d4ae2c84d8a096b4d39)
- #[serde(untagged)], i.e. untagged enums [1](https://github.com/ron-rs/ron#user-content-fn-serde-enum-hack-59fb02d3c6aa3d4ae2c84d8a096b4d39)
- #[serde(flatten)], i.e. flattening of structs into maps [2](https://github.com/ron-rs/ron#user-content-fn-serde-flatten-hack-59fb02d3c6aa3d4ae2c84d8a096b4d39)

in addition to anything using `deserialize_any` are not currently fully supported by eframe, but there's an easy way to support them by using e.g. JSON instead of RON (and there may be other better format choices that also support all serde features). So then I am curious why RON was chosen initially since I can't see the benefit over other formats. I mean, surely there _are_ benefits, right?

`serde` works, `eframe` works, `serde_json` works, but `ron` fails. If you can provide me with some motivation to try to improve the RON project, I might be willing to debug for them, but so far I can't see a good reason to put effort into that format.

**Reply by cjs-axsh** [reply]
I just read through my sdk a bit, and I don't think there is any deserialize_any usage in it, but there is quite a bit of enum in newtype in enum in newtype sort of relationships often tagged with the explicitly should-generally-work-labeled tags listed above. So then if this as all correct, there is maybe some other RON bug as you initially guessed. But I'm not sure and still not willing to become sure without some reason as to why RON is better for eframe than e.g. JSON.

**Reply by kvark** [reply]
Thank you for digging through this!
Reasons for RON against JSON are listed in the RON repo readme:
> Note the following advantages of RON over JSON:
- trailing commas allowed
- single- and multi-line comments
- field names aren't quoted, so it's less verbose
- optional struct names improve readability
- enums are supported (and less verbose than their JSON representation)

Obviously, if your use case requires some of the serde features are required that RON doesn't support right now, then it's more difficult to adopt. Some of these limitations are hopefully going to be fixed and just a subject to engineering effort put into the implementation.
I can't comment on why EFrame specifically chose RON but I'm glad it did - it's just a pleasure to read and write with Rust.

**Reply by cjs-axsh** [reply]
Everything you listed is in line with "enums are easier to write" as I initially stated and doesn't seem relevant to the eframe case as I mentioned in my opening comment.

> I'm glad it did - it's just a pleasure to read and write with Rust.

So you are actually manually reading and writing the persistence file by hand? This surprises me -- can you maybe explain why you need to do that? I feel like I must be missing something here.

**Reply by kvark** [reply]
I don't know how this list of features fares against eframe usage of RON. My comment was about RON in general, i.e. "why would X choose RON over JSON".

**Reply by cjs-axsh** [reply]
Does this mean that you don't actually edit the eframe persistence file by hand and therefore don't need to care how easy it is to read and write for humans?

**Reply by kvark** [reply]
Technically speaking, I don't need to care, and I'm not currently in business of editing eframe persistence files.

**Reply by cjs-axsh** [reply]
I see. So back to the drawing board for my question I guess. Thanks anyway.
